rust/lance-index/src/vector/hnsw/builder.rs

```diff
@@ -418,33 +426,36 @@ impl HnswBuilder {
                     current_node.add_neighbor(neighbor.id, neighbor.dist, level);
                 }
                 self.prune(storage, &mut current_node, level);
-                pruned_neighbors_per_level[level as usize]
-                    .clone_from(&current_node.level_neighbors_ranked[level as usize]);
+
+                for ordered_node in &current_node.level_neighbors_ranked[level as usize] {
+                    rev_edges.push(Edge {
+                        origin: ordered_node.id,
+                        level,
+                        distance: ordered_node.dist,
+                        destination: node,
+                    });
+                }
 
                 ep = neighbors[0].clone();
             }
         }
-        for (level, pruned_neighbors) in pruned_neighbors_per_level.iter().enumerate() {
-            let _: Vec<_> = pruned_neighbors
-                .iter()
-                .map(|unpruned_edge| {
-                    let level = level as u16;
-                    let m_max = match level {
-                        0 => self.params.m * 2,
-                        _ => self.params.m,
-                    };
-                    if unpruned_edge.dist
-                        < nodes[unpruned_edge.id as usize]
-                            .read()
-                            .unwrap()
-                            .cutoff(level, m_max)
-                    {
-                        let mut chosen_node = nodes[unpruned_edge.id as usize].write().unwrap();
-                        chosen_node.add_neighbor(node, unpruned_edge.dist, level);
-                        self.prune(storage, &mut chosen_node, level);
-                    }
-                })
-                .collect();
+        (target_level, rev_edges)
+    }
+
+    fn insert_backward(
+        &self,
+        node: u32,
+        unpruned_neighbors_per_level_rev: Vec<Vec<OrderedNode>>,
+        storage: &impl VectorStore,
+    ) {
+        let nodes = &self.nodes;
+        for (level, unpruned_neighbors) in unpruned_neighbors_per_level_rev.iter().enumerate() {
+            let mut current_node = nodes[node as usize].write().unwrap();
+            unpruned_neighbors.iter().for_each(|unpruned_edge| {
+                let level = level as u16;
+                current_node.add_neighbor(unpruned_edge.id, unpruned_edge.dist, level);
```

Comment:
maybe add a debug assertion here to assert `unpruned_edge.id != node`

Reply:
Added